snakemake/executors.py

```diff
@@ -2115,3 +2128,741 @@ def run_wrapper(
+
+        for job in self.active_jobs:
+            request = operations.cancel(name=job.jobname)
+            logger.debug("Cancelling operation {}".format(job.jobid))
+            try:
+                self._retry_request(request)
+            except (Exception, BaseException, googleapiclient.errors.HttpError):
+                continue
+
+        self.shutdown()
+
+    def _generate_job_resources(self, job):
+        """given a particular job, generate the resources that it needs,
+           including default regions and the virtual machine configuration
+        """
+        # Right now, do a best effort mapping of resources to instance types
+        cores = job.resources.get("_cores", 1)
+        mem_mb = job.resources.get("mem_mb", 100)
+        disk_mb = job.resources.get("disk_mb", 128000)
+
+        # Convert mb to gb, add buffer of 50
+        disk_gb = math.ceil(disk_mb / 1024) + 10
+
+        # Regular expression to determine if zone in region
+        regexp = "^(%s)" % "|".join(self.regions)
+
+        # Retrieve zones, filter down to selected regions
+        zones = self._retry_request(
+            self._compute_cli.zones().list(project=self.project)
+        )
+        zones = [z for z in zones["items"] if re.search(regexp, z["name"])]
+
+        # Retrieve machine types available across zones
+        # https://cloud.google.com/compute/docs/regions-zones/
+        lookup = {}
+        for zone in zones:
+            request = self._compute_cli.machineTypes().list(
+                project=self.project, zone=zone["name"]
+            )
+            lookup[zone["name"]] = self._retry_request(request)["items"]
+
+        # Only keep those that are shared, use last zone as a base
+        machine_types = {mt["name"]: mt for mt in lookup[zone["name"]]}
+        del lookup[zone["name"]]
+
+        # Update final list based on the remaining
+        to_remove = set()
+        for zone, types in lookup.items():
+            names = [x["name"] for x in types]
+            names = [name for name in names if "micro" not in name]
+            names = [name for name in names if not re.search("^(e2|m1)", name)]
+            for machine_type in list(machine_types.keys()):
+                if machine_type not in names:
+                    to_remove.add(machine_type)
+
+        for machine_type in to_remove:
+            del machine_types[machine_type]
+
+        # Alert the user of machine_types available before filtering
+        # https://cloud.google.com/compute/docs/machine-types
+        logger.debug(
+            "found {} machine types across regions {} before filtering "
+            "to increase selection, define fewer regions".format(
+                len(machine_types), self.regions
+            )
+        )
+
+        # First pass - eliminate anything that too low in cpu/memory
+        keepers = dict()
+
+        # Also keep track of max cpus and memory, in case none available
+        maxCpu = 1
+        maxMem = 100
```

Comment:
please always use snake-case: max_cpu, max_mem, ...

Reply:
I must have been distracted by the Google keys that use camel! Fixed now.